packages/amplify-opensearch-simulator/src/index.ts

```diff
@@ -0,0 +1,368 @@
+import waitPort from 'wait-port';
+import detectPort from 'detect-port';
+import execa from 'execa';
+import { ensureDir, writeFileSync, existsSync } from 'fs-extra';
+import gunzip from 'gunzip-maybe';
+import nodefetch from 'node-fetch';
+import { join } from 'path';
+import { pipeline, Readable } from 'stream';
+import tar from 'tar';
+import { promisify } from 'util';
+const { fromEvent } = require('promise-toolbox');
+import * as openpgp from 'openpgp';
+import { $TSAny, AmplifyFault, AMPLIFY_SUPPORT_DOCS, isWindowsPlatform, GetPackageAssetPaths, pathManager } from 'amplify-cli-core';
+import { printer } from 'amplify-prompts';
+
+// default port that opensearch chooses
+const basePort = 9200;
+
+const defaultOptions: Required<OpenSearchEmulatorOptions> = {
+  clusterName: 'opensearch-cluster',
+  nodeName: 'opensearch-node-local',
+  port: basePort,
+  type: 'single-node',
+  startTimeout: 20 * 1000,
+};
+
+const retryInterval = 20;
+const maxRetries = 5;
+export const supportedOpenSearchVersion = '1.3.0'; // latest version of OpenSearch supported by AWS ES
+export const relativePathToOpensearchLocal = join('opensearch', supportedOpenSearchVersion);
+export const packageName = 'amplify-opensearch-simulator';
+
+type OpenSearchEmulatorOptions = {
+  port?: number; 
+  clusterName?: string; 
+  nodeName?: string;
+  type?: string; 
+  startTimeout?: number;
+}
+
+export class OpenSearchEmulator {
+  public proc: execa.ExecaChildProcess<string>;
+  public opts: OpenSearchEmulatorOptions;
+
+  constructor(proc: execa.ExecaChildProcess<string>, opts: OpenSearchEmulatorOptions) {
+    this.proc = proc;
+    this.opts = opts;
+    return this;
+  }
+
+  public get pid(): number|undefined {
+    return this.proc?.pid;
+  }
+
+  public get port(): number|undefined {
+    return this.opts?.port;
+  }
+
+  public get url(): string {
+    return `http://localhost:${this.port}/`;
+  }
+
+  public terminate() {
```

Comment:
As best I can tell, this has a return type of the following:

```ts
  public terminate(): number | undefined | Promise<unknown> {
```

Please define the return types on all functions. If the above return type is accurate, a refactor is in order.

Reply:
I've tried to disambiguate types wherever possible. There are still couple of functions like this one which need a deep dive and refactor. Like I noted above, can we do this when we abstract out common components of DynamoDB and opensearch emulators?